text/3730-nit-lint-level.md

```diff
@@ -0,0 +1,479 @@
+worked with, a linter used a similar database to track what lints were "new" and reported those back to
+[ReviewBoard](https://www.reviewboard.org/).
+
+## Choosing lint levels
+
+When creating a lint or overriding a default lint level,
+its important to consider the semantics, something like:
+
+`error` is for hard errors that even prevent reporting of errors further in the build, running of tests, etc.
+This should be reserved for when there is no point building or testing anything further because of how broken the code is.
+
+`warn` generally has the semantics of a soft-error.
+Its non-blocking locally but will be blocked when the code is merged upstream.
+This should be used when eventual correctness or consistency is desired.
+
+`nit` is for coding suggestions for users that may or may not improve the code or may not need to be done yet.
+
+`allow` is for when there is no right answer (e.g. mutually exclusive lints), there are false positives, the lint is overly noisy, or the lint is of limited value.
+
+*Note:* This is for communicating the intent of this RFC and to get people started with this new feature and is not meant to precisely redefine the
+[lint levels](https://rustc-dev-guide.rust-lang.org/diagnostics.html#diagnostic-levels).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Rustc
+
+Rustc focuses on the mechanics of a new, visible lint level `nit`.
+
+Add support for the new lint:
+```rust
+#![nit(deprecated)]
+```
+```console
+$ rustc --nit deprecated ...
+```
+```console
+$ rustc -Ndeprecated ...
+```
+
+These will be rendered like other lint levels with a diagnostic level and coloring that conveys the non-blocking, helpful nature.
+For example, this could be rendered like a `note:` or `help:` diagnostic level.
+
+Rustc will have a dynamic lint group of all `nit`s, much like `warnings` is all lints at level `warn` at the time of processing, `-Anits`.
+
+## Clippy
+
+Like Rustc, Clippy would focus on the mechanism.
+Add support for the new lint:
+```rust
+#![nit(deprecated)]
+```
+```console
+$ rustc --nit clippy::pedantic ...
+```
+```console
+$ rustc -Nclippy::pedantic ...
+```
+
+These will be rendered like Rustc.
+
+These will be part of Rustc's dynamic lint group, much like `warnings`.
+
+## Cargo
+
+Cargo implements the semantics for this to be a first-class non-blocking lint level.
+
+Add support for the new lint:
+```toml
+[workspace.lints.rust]
+deprecated = "nit"
+```
+
+A new config field will be added to mirror `build.warnings`:
+
+### `build.nits`
+
+- Type: string
+- Default: `allow`
+- Environment: `CARGO_BUILD_NITS`
+
+Controls how Cargo handles nits. Allowed values are:
+
+- `nit`: warnings are emitted as warnings.
+- `allow`: warnings are hidden (default).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Rust-analyzer might be limited in what it can do through LSP to avoid overwhelming the user with `nit`s.
```

Comment:
I don't like nits being left to IDEs. 

Rust-Analyzer/LSP are very biased towards MS Visual Studio Code. Where LSP doesn't support something, Rust Analyzer uses VSCode-specific workarounds, but other editors are left out. Other editors usually have sub-par experience. Sublime Text for example loses a lot of fidelity when displaying Rust's errors (in most cases it can't show a connection between a span and a note).

The problem is that Rust Analyzer uses VS Code as the reference implementation of LSP, and doesn't want to test and work around issues in other editors. Developers of the other IDEs/editors also just want to support LSP, and don't want to add features/workarounds for Rust Analyzer specifically. This means that any issues with how Rust errors are displayed in an LSP IDE end up being nobody's responsibility: RA tells to file bugs with your editor, and the editor says to file bugs with RA.

Reply:
Yeah, I don't use Rust Analyzer, due to a bunch of different reasons: I've historically always worked on larger codebases, often with custom build systems, and IDEs have never been great at those. I also use Sublime Text and Rust IDE support has always lagged there. Both these problems are not as big a deal as they used to be now, but at this point ten years of writing Rust have made my existing workflows comparably efficient to using an IDE and I don't wish to put effort into switching now.